### ISSUE TITLE
[css-backgrounds-4] Add background-repeat-x/y

### ISSUE THREAD

**Issue by mstange** [w3c/csswg-drafts#116]
It looks like recent versions of Edge and Chrome support the properties background-repeat-x and background-repeat-y. I haven't checked how far back support goes. Gecko does not support them, and Webkit doesn't appear to either.
If this is something people want, then it should be added to the spec.

**Comment by yisibl** [comment]
Agree, I think it should be consistent with background-position-x/y.

**Comment by fantasai** [comment]
What's the use case for cascading them independently?

**Comment by SebastianZ** [comment]
Backgrounds Level 4 already [defines those two properties plus `background-position-inline` and `background-position-block`](https://drafts.csswg.org/css-backgrounds-4/#background-position-longhands). So, I'm wondering why this issue was opened.

> Gecko does not support them, …

That info is outdated. They were [implemented in Gecko 49](https://bugzilla.mozilla.org/show_bug.cgi?id=550426).

> … and Webkit doesn't appear to either.

I don't have the means to test it, but the related MDN articles suggest that WebKit does support [`background-position-x`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position-x) and [`background-position-y`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position-y). Could someone please verify that and update the articles accordingly (or let me know about it, so I can do that)?

Sebastian

**Comment by mstange** [comment]
Sebastian, this is about background-*repeat*, not background-position. (Also, I'm the person who implemented background-position-x/y in Gecko.)

Fantasai, I don't have a use case. Maybe it's just about consistency with background-position? To me, background-repeat "feels" like a shorthand property because it allows you to specify different values for each axis.

**Comment by fantasai** [comment]
If there's no use case for cascading them independently, then we should not add longhands. And Chrome and Edge should remove them.

**Comment by tabatkins** [comment]
[Our usage numbers](https://www.chromestatus.com/metrics/css/popularity#background-repeat-y) show it at .05%, above our standard removal threshold.

**Comment by SebastianZ** [comment]
> Sebastian, this is about background-_repeat_, not background-position. (Also, I'm the person who implemented background-position-x/y in Gecko.)

Oops! Please forgive me my blindness! I saw you've also already filed [bug 1273244](https://bugzilla.mozilla.org/show_bug.cgi?id=1273244) for it.

Sebastian

**Comment by dbaron** [comment]
I wrote [a few tests](https://dbaron.org/css/test/2017/background-repeat-x-y) to allow testing which implementations support these.

**Comment by gregwhitworth** [comment]
Updating here that Edge does not support the longhands.

**Comment by dbaron** [comment]
It turns out that the [group resolved to add these in April 2014](https://lists.w3.org/Archives/Public/www-style/2014Apr/0188.html).

(We also re-discussed them today, twice.)

**Comment by gregwhitworth** [comment]
After discussing this after the resolution, I wasn't aware that deprecated implied that all other UAs that don't support this need to add this - my understanding of it's implication was more inline with obsolete. I would like to re-open this issue and discuss roughly defining it but marking it as obsolete. In addition to this, @FremyCompany looked into why Edge removed this and it was due to us implementing multiple backgrounds, and when testing Chrome sure enough you can't have multiple backgrounds in conjunction with background-repeat-x/y (https://jsfiddle.net/u5L8ak8b/2/ - I tried numerous vals to see if they mapped them together to no avail). By switching to obsolete, it allows us to vocalize to authors not to use it and allows Blink to, when possible take the necessary steps to remove it and doesn't burden others to implement it.

**Comment by KuoE0** [comment]
@gregwhitworth I think your test case missing a comma between two urls in `background-image`. And it's true that `background-repeat-{x,y}` can not accept multiple values. So you can only choose one combination for the two images in Blink. It means you can't create the same result created by the following source code with `background-repeat-{x,y}`.
```
background-repeat: repeat no-repeat, no-repeat repeat;
```

**Comment by gregwhitworth** [comment]
@KuoE0 thanks for catching the missing comma. Even with that, it still seems odd to have a long hand that doesn't support that of the shorthand and so I still stand by allowing it in the spec and marked as obsolete, this allows Blink to, when they deem it possible, to remove it, but also doesn't imply that other UAs that don't support it to implement it.

**Comment by dbaron** [comment]
So at the face-to-face meeting we [discussed this once](https://log.csswg.org/irc.w3.org/css/2017-01-11/#e758676) and then later in the afternoon [discussed it again](https://log.csswg.org/irc.w3.org/css/2017-01-11/#e759036).  Despite that we did reach a conclusion the second time, it seems that given  Greg's https://github.com/w3c/csswg-drafts/issues/116#issuecomment-272232037 above, I think we need to discuss it again.

**Comment by tantek** [comment]
@gregwhitworth I should have said obsolete not deprecate at the f2f. I misspoke :(

**Comment by dbaron** [comment]
We [just discussed this again](https://log.csswg.org/irc.w3.org/css/2017-01-25/#e766398) at the teleconference, ending with:

> RESOLVED: Do not spec background-repeat-x/y for now

**Comment by SebastianZ** [comment]
According to [Google's data and the data from HTTP Archive](https://chromestatus.com/metrics/css/timeline/popularity/31), the use of `background-repeat-x` significantly increased over the past years and is now at 0.485%. The usage of `background-repeat-y` didn't increase as much but is still relatively high with 0.2%.

This indicates that there _is_ some need for those longhands, even when there is only limited support by a single browser engine, at the moment.
Maybe someone familiar with querying HTTP Archive could find out for what pages are currently using those properties.

Sebastian

**Comment by gsnedders** [comment]
There's some confusion above about whether WebKit does support them or not: it… sorta does—and Blink got it from WebKit. (This dates back to the original implementation of two-value `background-repeat` in WebKit.) The confusion likely arrises because we avoid exposing it… sometimes.

FWIW, I tried to remove these from WebKit as part of https://github.com/WebKit/WebKit/commit/d70cd13f28142083ebffc879426101ad5a75e0d9 (given they were in many common cases _not_ exposed)… and I've already had some breakage reported (because JS was assuming there were at least 10 longhand properties of `background`!).

Not made any decision about what we're going to do, but I strongly suspect we won't go back to the "semi-exposed" behaviour we had before, which means either totally shipping it or totally removing it.

See also https://bugs.chromium.org/p/chromium/issues/detail?id=1073573 for related oddness in Blink.

**Comment by SebastianZ** [comment]
Thanks @gsnedders for the implementation insights!

Quoting my own comment

> According to [Google's data and the data from HTTP Archive](https://chromestatus.com/metrics/css/timeline/popularity/31), the use of `background-repeat-x` significantly increased...
> 
> This indicates that there _is_ some need for those longhands...
> Maybe someone familiar with querying HTTP Archive could find out for what pages are currently using those properties.

The goal behind this request was to get an answer for @fantasai's [question back in 2016](https://github.com/w3c/csswg-drafts/issues/116#issuecomment-221479056), i.e. to find out whether there _are_ actually use cases for those two being defined independently.

I assume there is a use case for specifying `no-repeat` for one axis while keeping the other axis unchanged, i.e. `repeat`.
So defining `background-image: url('pattern.png'); background-repeat-y: no-repeat;` would result in a decorative strip. And an author wouldn't have to write `background-repeat: repeat no-repeat;`.

Skimming through the previous discussions, I still wonder what's the issue with those longhands that lead to the resolution to make them deprecated and lead implementers to remove them rather than specify and implement them properly.
(Note that the above is meant objectively. I don't have a strong opinion for or against adding them.)

Sebastian

**Comment by SebastianZ** [comment]
Reiterating on my previous comments, the usage numbers of the two properties are steadily increasing. According to Google's data, we're now at ~0.513% and 0.241%, respectively.

This looks like a strong indicator that those longhands should be specified and implemented properly after all.

@gsnedders @dbaron Maybe one of you can answer the question why it was previously decided that they should be removed.

Sebastian

**Comment by SebastianZ** [comment]
Bringing this to the table again because the usage numbers are still increasing.

The proposal is to bring `background-repeat-x` and `background-repeat-y` back and properly define them.

Sebastian

**Comment by Loirooriol** [comment]
Just want to note that splitting the property may make it hard (since we don't have a solution for #1282) to add new logical values like `background-repeat: repeat-block`. But of course, we might never add such values, or it will be fine when we solve #1282.

**Comment by Loirooriol** [comment]
I'm opposed to both turning it into a shorthand and adding logical values with no plan about how to handle that. I'm fine with turning it into a shorthand without the logical values, or adding the logical values without turning it into a shorthand.

Same problem as https://github.com/w3c/csswg-drafts/issues/549#issuecomment-1856378331

**Comment by SebastianZ** [comment]
@Loirooriol The situation here seems to be a different one. `background-repeat` already has physical based definition for its two-value syntax and supports `repeat-x` and `repeat-y` as keywords.
The logical keywords `repeat-block` and `repeat-inline` are added to complement the existing keywords and there isn't a conflict in `var()` usage or CSSOM.

Just in case we add a way to change how the two-value syntax should be interpreted, we run into the issues you refer to. Or is there something I am missing?

(Note that I turned `background-repeat` into a shorthand in #9756 and added the logical keywords for it now in #10755.)

Sebastian

**Comment by Loirooriol** [comment]
So basically you want `background-repeat: repeat-block` to set `background-repeat-x` and `background-repeat-y` to an internal value that won't be resolved into used (why not computed?) value time?

I think people would generally expect it to expand into `background-repeat-block: repeat; background-repeat-inline: no-repeat`.

**Comment by cdoublev** [comment]
Is there any reason not to define a [logical property group](https://drafts.csswg.org/css-logical-1/#logical-property-group) for `background-repeat-*` longhands? That is, I expect `background-repeat` to be defined as a `Logical property group` in the property definition table.

This would be needed *to maintain relative order of declarations within a CSS declaration block*, eg. with `style.cssText = 'background-repeat: repeat; background-repeat-block: no-repeat; background-repeat-y: repeat'`.

**Comment by cdoublev** [comment]
Both `background-repeat` and [`mask-repeat`](https://drafts.fxtf.org/css-masking-1/#propdef-mask-repeat) value definitions include `<repeat-style>`. CSS Backgrounds 4 now allows two keywords. It may be ok as long as CSS Masks 1 links to the single keyword definition in CSS Backgrounds 3, but I think either one of them should use a separate definition, or `mask-repeat-*` should be specced like `background-repeat`.

Can you please clarify the CSSWG's position on `mask-repeat`? [w3c/fxtf-drafts#99](https://github.com/w3c/fxtf-drafts/issues/99) confuses me because I do not interpret the related [discussions/resolution](https://github.com/w3c/csswg-drafts/issues/116#issuecomment-275179463) (from this issue) as in the comment that closed it.

**Comment by SebastianZ** [comment]
@Loirooriol:

> So basically you want `background-repeat: repeat-block` to set `background-repeat-x` and `background-repeat-y` to an internal value that won't be resolved into used (why not computed?) value time?
> 
> I think people would generally expect it to expand into `background-repeat-block: repeat; background-repeat-inline: no-repeat`.

You're right. I updated my PR to reflect your remark. Let me know if the changes makes sense to you.

Sebastian

**Comment by Loirooriol** [comment]
> I updated my PR to reflect your remark. Let me know if the changes makes sense to you.

I don't see my remark addressed.

**Comment by cdoublev** [comment]
+1. I do not know how to expand `background[-repeat]: repeat-block`.

**Comment by SebastianZ** [comment]
The spec now says

> `repeat-block`
>    Computes to [`repeat`](https://drafts.csswg.org/css-backgrounds-4/#valdef-background-repeat-x-repeat) in the [block axis](https://drafts.csswg.org/css-writing-modes-4/#block-axis) and [`no-repeat`](https://drafts.csswg.org/css-backgrounds-4/#valdef-background-repeat-x-no-repeat) in the [inline axis](https://drafts.csswg.org/css-writing-modes-4/#inline-axis). 
repeat-inline
>    Computes to [`repeat`](https://drafts.csswg.org/css-backgrounds-4/#valdef-background-repeat-x-repeat) in the [inline axis](https://drafts.csswg.org/css-writing-modes-4/#inline-axis) and [`no-repeat`](https://drafts.csswg.org/css-backgrounds-4/#valdef-background-repeat-x-no-repeat) in the [block axis](https://drafts.csswg.org/css-writing-modes-4/#block-axis).

and @cdoublev [made the longhands form a logical property group](https://github.com/w3c/csswg-drafts/pull/11139).

@Loirooriol @cdoublev Could you please specify how that can be further clarified?

Sebastian

**Comment by Loirooriol** [comment]
Either the shorthand expands to the physical longhands only, and then logical values need to be represented as an internal pending value at parse time, or the shorthand expands to logical or physical longhands depending on the provided value.

The 2nd option is what most people would probably expect, but has lots of problems (in particular, it breaks with variables).

Thus my opposition in https://github.com/w3c/csswg-drafts/issues/116#issuecomment-1856386111. People promised some magical solution that I still have not seen anywhere.

**Comment by cdoublev** [comment]
Shorthands of sub-properties with different mapping logics are still under discussions, indeed.

I raised other related questions in #9253 for `background-position`, in which I think I assumed option 2, but I am not sure I considered the pending substitution value issue.
